src/main/scala/io/iohk/ethereum/network/p2p/Message.scala

```diff
@@ -60,44 +64,427 @@ object Hello {
+      case rlpList: RLPList => GetBlockBodies(rlpList.items.map(e => ByteString(e: Array[Byte])))
+
+      case _ => throw new RuntimeException("Cannot decode BlockHeaders")
+    }
+  }
+
+  val code: Int = 0x10 + 0x05
 }
 
+case class GetBlockBodies(hashes: Seq[ByteString]) extends Message {
+  override def code: Int = GetBlockBodies.code
+
+  override def toString: String = {
+    s"""GetBlockBodies {
+       |hashes: ${hashes.map(h => Hex.toHexString(h.toArray[Byte]))}
+       |}
+     """.stripMargin
+  }
+}
+
+object Transactions {
+  implicit val rlpEndDec = new RLPEncoder[Transactions] with RLPDecoder[Transactions] {
+    override def encode(obj: Transactions): RLPEncodeable = {
+      import obj._
+      RLPList(txs.map(Transaction.rlpEndDec.encode): _*)
+    }
+
+    override def decode(rlp: RLPEncodeable): Transactions = rlp match {
+      case rlpList: RLPList => Transactions(rlpList.items.map(Transaction.rlpEndDec.decode))
+      case _ => throw new RuntimeException("Cannot decode Transactions")
+    }
+  }
+
+  val code: Int = 0x10 + 0x02
+}
+
+case class Transactions(txs: Seq[Transaction]) extends Message {
+  override def code: Int = Transactions.code
+}
+
+object Transaction {
+  implicit val rlpEndDec = new RLPEncoder[Transaction] with RLPDecoder[Transaction] {
+    override def encode(obj: Transaction): RLPEncodeable = {
+      import obj._
+      RLPList(nonce, gasPrice, gasLimit, receivingAddress.toArray[Byte], value,
+        payload.fold(_.byteString.toArray[Byte], _.byteString.toArray[Byte]),
+        pointSign, signatureRandom.toArray[Byte], signature.toArray[Byte])
+    }
+
+    override def decode(rlp: RLPEncodeable): Transaction = rlp match {
+      case RLPList(nonce, gasPrice, gasLimit, (receivingAddress: RLPValue), value, payload, pointSign, signatureRandom, signature)
+        if receivingAddress.bytes.nonEmpty =>
+        Transaction(nonce, gasPrice, gasLimit, ByteString(receivingAddress: Array[Byte]), value, Right(TransactionData(ByteString(payload: Array[Byte]))),
+          pointSign, ByteString(signatureRandom: Array[Byte]), ByteString(signature: Array[Byte]))
+
+      case RLPList(nonce, gasPrice, gasLimit, (receivingAddress: RLPValue), value, payload, pointSign, signatureRandom, signature)
+        if receivingAddress.bytes.isEmpty =>
+        Transaction(nonce, gasPrice, gasLimit, ByteString(), value, Left(ContractInit(ByteString(payload: Array[Byte]))),
+          pointSign, ByteString(signatureRandom: Array[Byte]), ByteString(signature: Array[Byte]))
+
+      case _ => throw new RuntimeException("Cannot decode Transaction")
+    }
+  }
+}
+
+//ETH yellow paper section 4.3
+case class Transaction(
+  nonce: BigInt,
+  gasPrice: BigInt,
+  gasLimit: BigInt,
+  receivingAddress: ByteString,
+  value: BigInt,
+  payload: Either[ContractInit, TransactionData],
+  //yellow paper appendix F
+  pointSign: Byte, //v - 27 or 28 according to yellow paper, but it is 37 and 38 in ETH
```

Comment:
it does, I think I will implement it in next iteration, because we have to address situation when sign is 37/38 instead of 27/28 which happens in some transactions